src/state_oorset_ps.erl

```diff
@@ -0,0 +1,890 @@
+
+-define(TYPE, ?MODULE).
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-export([new/0, new/1]).
+-export([mutate/3, delta_mutate/3, merge/2]).
+-export([query/1, equal/2, is_bottom/1, is_inflation/2, is_strict_inflation/2]).
+-export([join_decomposition/1]).
+
+-export([subtract_all_events/2,
+         add_elem_with_dot/3,
+         get_events_from_provenance/1,
+         cross_provenance/2,
+         subtract_removed/2]).
+
+-export_type([state_oorset_ps/0,
+              delta_state_oorset_ps/0,
+              state_oorset_ps_op/0]).
+
+-opaque state_oorset_ps() :: {?TYPE, payload()}.
+-opaque delta_state_oorset_ps() :: {?TYPE, {delta, payload()}}.
+-type delta_or_state() :: state_oorset_ps() | delta_state_oorset_ps().
+-type payload() :: {ps_data_store(), ps_filtered_out_events(), ps_all_events()}.
+-type element() :: term().
+-type state_oorset_ps_op() :: {add, element()}
+                               | {add_all, [element()]}
+                               | {rmv, element()}
+                               | {rmv_all, [element()]}.
+
+%% Provenanve semiring related.
+-export_type([ps_event_id/0]).
+
+-type ps_object_id() :: binary().
+-type ps_replica_id() :: term().
+-type ps_event_id() :: {ps_object_id(), ps_replica_id()}.
+-type ps_event() :: {ps_event_id(), pos_integer()}.
+-type ps_dot() :: ordsets:ordsets(ps_event()).
+-type ps_provenance() :: ordsets:ordsets(ps_dot()).
+-type ps_data_store() :: {orddict:orddict(element(), ps_provenance()),
+                          orddict:orddict(ps_event(), element())}.
+-type ps_filtered_out_events() :: ps_dot().
+-type ps_all_events() :: ps_dot()                %% for delta state
+                       | {vclock, [ps_event()]}. %% for full state
+
+%% @doc Create a new, empty `state_oorset_ps()'
+-spec new() -> state_oorset_ps().
+new() ->
+    {?TYPE, {{orddict:new(), orddict:new()}, ordsets:new(), {vclock, []}}}.
+
+%% @doc Create a new, empty `state_oorset_ps()'
+-spec new([term()]) -> state_oorset_ps().
+new([]) ->
+    new().
+
+%% @doc Mutate a `state_oorset_ps()'.
+-spec mutate(state_oorset_ps_op(), type:id(), state_oorset_ps()) ->
+    {ok, state_oorset_ps()} | {error, {precondition, {not_present, [element()]}}}.
+mutate(Op, Actor, {?TYPE, _ORSet}=CRDT) ->
+    state_type:mutate(Op, Actor, CRDT).
+
+%% @doc Delta-mutate a `state_oorset_ps()'.
+%%      The first argument can be:
+%%          - `{add, element()}'
+%%          - `{add_all, [element()]}'
+%%          - `{rmv, element()}'
+%%      The second argument is the event id ({object_id, replica_id}).
+%%      The third argument is the `state_oorset_ps()' to be inflated.
+-spec delta_mutate(state_oorset_ps_op(), type:id(), state_oorset_ps()) ->
+    {ok, delta_state_oorset_ps()} | {error, {precondition, {not_present, element()}}}.
+%% Adds a single element to `state_oorset_ps()'.
+%% Delta: {{[{Elem, {{NewEvent}}}], [{NewEvent, Elem}]}, [], [NewEvent]}
+delta_mutate({add, Elem},
+             EventId,
```

Comment:
Yes, 'EventId' should be unique. In order to handle the union()/intersection()/product() operations properly, the actor identifier was not enough. Therefore, I created the EventId which is {ObjectId, ActorId}. If this is not allowed, then I have to find another way for it since this is an essential requirement for those operations.